File: trunk/proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/reports/DocumentIndex.java
package com.jasonzqshen.familyaccounting.core.reports;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Hashtable;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.listeners.LoadDocumentListener;
import com.jasonzqshen.familyaccounting.core.listeners.SaveDocumentListener;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataIdentity;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataManagement;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;

public abstract class DocumentIndex {
    /**
     * comparator based on posting date
     */
    public final static Comparator<HeadEntity> COMPARATOR_DATE = new Comparator<HeadEntity>() {
        @Override
        public int compare(HeadEntity head1, HeadEntity head2) {
            return head1.getPostingDate().compareTo(head2.getPostingDate());
        }
    };

    public final static int ACCOUNT_INDEX = 0;

    public final static int BUSINESS_INDEX = 1;

    public final static int INDEX_COUNT = 2;

    protected final Hashtable<MasterDataIdentity, DocumentIndexItem> _list;

    protected final CoreDriver _coreDriver;

    protected final MasterDataManagement _mdMgmt;

    // load document listener
    private final LoadDocumentListener _loadDocumentListener = new LoadDocumentListener() {
        public void onLoadDocumentListener(Object source, HeadEntity document) {
            newDoc(document);
        }
    };

    // save document listener
    private final SaveDocumentListener _saveDocumentListener = new SaveDocumentListener() {
        public void onSaveDocumentListener(HeadEntity document) {
            newDoc(document);
        }
    };

    /**
     * document index
     * 
     * @param coreDriver
     */
    protected DocumentIndex(CoreDriver coreDriver, MasterDataManagement mdMgmt) {
        _coreDriver = coreDriver;
        _mdMgmt = mdMgmt;

        _list = new Hashtable<MasterDataIdentity, DocumentIndexItem>();

        _coreDriver.getListenersManagement().addSaveDocListener(
                _saveDocumentListener);
        _coreDriver.getListenersManagement().addLoadDocListener(
                _loadDocumentListener);
    }

    /**
     * 
     * @return
     */
    public ArrayList<MasterDataIdentity> getKeys() {
        ArrayList<MasterDataIdentity> ret = new ArrayList<MasterDataIdentity>(
                _list.keySet());
        Collections.sort(ret);

        return ret;
    }

    /**
     * get index item
     * 
     * @param key
     * @return
     */
    public DocumentIndexItem getIndexItem(MasterDataIdentity key) {
        return _list.get(key);
    }

    /**
     * clear
     */
    public void clear() {
        _list.clear();
    }

    /**
     * set report when new document
     * 
     * @param head
     */
    protected abstract void newDoc(HeadEntity head);
}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Reports/DocumentIndex.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Transaction;

namespace WinStore_FamilyAccountingCore.Reports
{
    public abstract class DocumentIndex
    {
        /// <summary>
        /// compararion
        /// </summary>
        public readonly static Comparison<HeadEntity> COMPARATOR_DATE = new Comparison<HeadEntity>(
            delegate(HeadEntity head1, HeadEntity head2)
            {
                return head1.PstDate.CompareTo(head2.PstDate);
            });

        public readonly static int ACCOUNT_INDEX = 0;

        public readonly static int BUSINESS_INDEX = 1;

        public readonly static int INDEX_COUNT = 2;

        protected readonly Dictionary<MasterDataIdentity, DocumentIndexItem> _list;

        protected readonly CoreDriver _coreDriver;

        protected readonly MasterDataManagement _mdMgmt;


        // load document listener
        private void documentLoad(Object source, HeadEntity document)
        {
            newDoc(document);
        }

        // save document listener
        private void documentSaved(HeadEntity document)
        {
            newDoc(document);
        }


        // reveser document listener

        private void documentReversed(HeadEntity doc)
        {
            reverseDoc(doc);
        }

        /// <summary>
        /// document index
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="mdMgmt"></param>
        protected DocumentIndex(CoreDriver coreDriver, MasterDataManagement mdMgmt)
        {
            _coreDriver = coreDriver;
            _mdMgmt = mdMgmt;

            _list = new Dictionary<MasterDataIdentity, DocumentIndexItem>();

            _coreDriver.ListenerMgmt.DocumentSavedHandler += documentSaved;
            _coreDriver.ListenerMgmt.DocumentLoadHandler += documentLoad;
            _coreDriver.ListenerMgmt.DocumentReversedHandler += documentReversed;
        }

        /// <summary>
        /// keys
        /// </summary>
        public List<MasterDataIdentity> Keys
        {
            get
            {
                List<MasterDataIdentity> ret = new List<MasterDataIdentity>(
                        _list.Keys);
                ret.Sort();

                return ret;
            }
        }

        /// <summary>
        /// get index item
        /// </summary>
        /// <param name="key"></param>
        /// <returns></returns>
        public DocumentIndexItem getIndexItem(MasterDataIdentity key)
        {
            DocumentIndexItem item;
            if (_list.TryGetValue(key, out item))
            {
                return item;
            } 
            return null;
        }

        /// <summary>
        /// clear
        /// </summary>
        public void Clear()
        {
            _list.Clear();
        }

        /// <summary>
        /// set report when new document
        /// </summary>
        /// <param name="head"></param>
        protected abstract void newDoc(HeadEntity head);

        protected abstract void reverseDoc(HeadEntity head);
    }
}
